spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java

```diff
@@ -173,11 +172,13 @@ public void replaceChannelMappings(Properties channelMappings) {
 	}
 
 	private void doSetChannelMappings(Map<String, String> newChannelMappings) {
-		Map<String, String> oldChannelMappings = this.channelMappings;
-		this.channelMappings = newChannelMappings;
+		Map<String, String> oldChannelMappings = new HashMap<String, String>(this.channelMappings);
+		synchronized (this.channelMappings) {
```

Comment:
:+1: and even during iteration we end up with the same `ConcurrentModificationException` :smile: 

Any other concerns to push the valuable commit ?